### PR TITLE
Chore: Refactor Plan definition into immutable Plan and mutable PlanBuilder

### PR DESCRIPTION
**Pull request by izeigerman**

### CODE REVIEW

File: sqlmesh/core/plan/builder.py [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+        if not self._start and is_dev and forward_only:
+            self._start = default_start or yesterday_ds()
+
+        self._environment_naming_info = EnvironmentNamingInfo.from_environment_catalog_mapping(
+            environment_catalog_mapping or {},
+            name=self._context_diff.environment,
+            suffix_target=environment_suffix_target,
+        )
+
+        self._override_start = start is not None
```
Comment (tobymao) [inline sqlmesh/core/plan/builder.py:109]: just make these public and delete the properties

File: sqlmesh/core/plan/builder.py [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+        self._override_end = end is not None
+        self._plan_id: str = random_id()
+        self._model_fqn_to_snapshot = {s.name: s for s in self._context_diff.snapshots.values()}
```
Comment (tobymao) [inline sqlmesh/core/plan/builder.py:112]: isn't this exactly the same as context_diff.snapshots?

Comment (tobymao) [reply sqlmesh/core/plan/builder.py:112]: ah i guess snapshots is snapshot_id

File: sqlmesh/core/plan/builder.py (hunk 1 above)
Comment (izeigerman) [reply sqlmesh/core/plan/builder.py:109]: This is intentional to signal that these values are not supposed to be modified

**Review by tobymao — APPROVED** [review]

File: sqlmesh/core/context.py [hunk 3]
```diff
@@ -826,6 +826,88 @@ def plan(
...
+            run=run,
+        )
+
+        self.console.plan(
+            plan_builder, auto_apply, self.default_catalog, no_diff=no_diff, no_prompts=no_prompts
+        )
+
+        return plan_builder.build()
+
+    def plan_builder(
```
Comment (eakmanrq) [inline sqlmesh/core/context.py:855]: Sure you want to make this public? Not sure if we want to commit to exposing this.

**Review by eakmanrq — APPROVED** [review]
Excited for this!

File: sqlmesh/core/plan/builder.py (hunk 1 above)
Comment (tobymao) [reply sqlmesh/core/plan/builder.py:109]: meh, no one's gonna modify it

Comment (izeigerman) [reply sqlmesh/core/plan/builder.py:109]: famous last words lol. Ok will change this

Comment (z3z1ma) [reply sqlmesh/core/plan/builder.py:109]: Interesting 👀

File: sqlmesh/core/context.py (hunk 3 above)
Comment (izeigerman) [reply sqlmesh/core/context.py:859]: Discussed in person. We concluded that this method makes a lot more sense when using SQLMesh API as opposed to the `plan` one which is clearly geared towards using in CLI or Notebook.

**Pull request merged**